### ISSUE TITLE
dotnet config

### ISSUE THREAD

**Issue by richlander** [dotnet/cli#10422]
It happened! We just past that moment where we absolutely need `dotnet config`. It's time to go build it, I propose for 3.0.

Git has `git config` as documented at [git config](https://git-scm.com/docs/git-config). We need something kinda/sorta the same.

Here are the things I'd like to consider setting for `dotnet config`:

- [bool] Use previews by default
- [string] Use this SDK version (same concept as global.json)
- [string] Use this `--framework` as the default for `dotnet new` (for example, `netcoreapp2.2`). This one might need both a default and a way to say "Oh, I want `netstandard2.0` for the `classlib` template).
- [string] Use this `--framework` as the default for `dotnet run` -- useful for multi-targeted projects.
- [string] Use this `--configuration` by default (for example `release`, `debug`)
- [array of strings] Use these flags for `dotnet test` by default (like the logger)
- [array of strings] Use these flags for `dotnet publish` by default (like self-contained of cross-gen)
- [array of strings] Always build for these RIDs
- [array of key/values] Environment variables to set at `dotnet run` process launch

global.json would still override these settings, where these is overlap.

**Comment by NickCraver** [comment]
I'm unclear how `--framework` specifically would work since we'd be configuring possibly outside the subset of available options for any given project (or even projects in the same solution). What happens when it's not one of the values? Or when it's just 1 of 3? Seems like that could get confusing and needs a lot of thought. The rest looks very appealing though, especially the global SDK defaults and preview control.

Looking forward to this!

**Comment by richlander** [comment]
> I'm unclear how `--framework` specifically would work since ...

Yes, I thought of that. `--framework` for `dotnet new` is a slam dunk. It is harder for `run` for exactly the reasons you raise. I wanted to offer a range of ideas and see if there are people that have a lot of projects that have the same frameworks specified. This idea would be a benefit for that case. I doubt we'd implement this particular idea in the first round.

Keep the feedback coming!

**Comment by normj** [comment]
Like the idea of this feature. Here are some suggestions to add to the config list.

* Ability to configure additional package store directories. I think custom package stores are going to become a big deal for us.
* Ability to configure the .NET Global Tools install folder. This could be useful in a build environment where a build script could set this to a folder in the workspace. Then the build script could install a tool without worry of making a change that could affect other build scripts. I would also have helped me during tool development not affecting my normal non-tool development.
* Ability to set the RID for the machine. This could be helpful for distributions that don't report their RID in the standard way .NET Core looks for. This is the case for Amazon Linux.

**Comment by richlander** [comment]
@normj -- are you referring to `--tool-path` for your second path?

> [array of strings] Always build for these RIDs

This is the same as your third point, yes?

**Comment by normj** [comment]
Startup hook looks really interesting and I need more time to digest it but I don't see how it would be able to adjust the **additional probing path** for the started process when the hook has been called and I don't want to get in the business of recreating the .NET Core dependency look up path.

I had forgot about `--tool-path`, that does solve that config need.

What I was looking for in the config for RID was not telling what RID to build for but override the .NET Core logic to determine what the current OS's RID is and use this config setting.

**Comment by livarcocc** [comment]
cc @nguerrera @johnbeisner 

It was considered for 3.0 but considered below the cut line by the directors. Given all the work that we already have for the release, I don't believe we can properly fund it.

Also, I believe @KathleenDollard already did a write up for this somewhere. Maybe she can share that here.

**Comment by onovotny** [comment]
One initial thought 
- would `dotnet config` be for the user profile or for a project/directory? If for the user profile, then my concern is differences between machines (works on my machine, breaks elsewhere, etc). While global.json would override, for some things that seems too subtle.

**Comment by onovotny** [comment]
I think one difference between git and dotnet is one affects how code is checked out/in, while the other affects build.

I think it's okay to put options for default new templates at the user/system level, but I would be worried about putting SDK/preview things at that level for fear of different builds per machine. I think certain settings should only be repo level, the ones that affect build/runtime behavior.

**Comment by nguerrera** [comment]
We could think about certain options being repo-only, but that feels too rigid to me. I think I'd prefer a way to opt out of any global configuration in the repo configuration. 

Returning to this:
> That said, one thing that significantly impacts the design and implementation challenge here is that we need to read the SDK selection options in native code before any managed code in the SDK can run

I just want to make it clear that the first two basic scenarios in this issue description are impacted by that. The code to honour those needs to live in native code in dotnet/core-setup.

**Comment by wli3** [comment]
> but that feels too rigid to me. I think I'd prefer a way to opt out of any global configuration in the repo configuration.

One example of this scheme is requiring signing for packages used. It should be a global flag. And you can optout per repo (although this config is in nuget)

**Comment by swaroop-sridhar** [comment]
>> configure additional package store directories ...
>> Is startup hook an option to replace that?

My understanding is that @normj is only asking for a new way to use an existing feature. 
That is, the CLI already supports `additionalProbingPaths` in `rutimeconfig.json` and `dotnet exec –additionalProbingPaths …` switch. The request is to add `dotnet config additionalProbingPaths`. Seems like a reasonable option to add to `dotnet config`

The use case described by @normj  is simple -- the additional directories to probe is known statically before the app starts, and doesn’t change. If we learn more complex/dynamic use-cases, we can provide samples for solving them using more powerful tools like [startup-hook](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-startup-hook.md), [AssemblyLoadContexts](https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/assemblyloadcontext.md), [NativeLibrary APIs](https://github.com/dotnet/corefx/issues/32015), etc.

**Comment by vitek-karas** [comment]
There's a related discussion in dotnet/core-setup#4809.
That one is about VS using the dotnet.exe from global location always - so we could use `global.json` (and/or the higher lever configs) to tell dotnet.exe where to look for frameworks/SDKs.

Regarding the need for this to be implemented in native code: yes that is absolutely true, most of these options would affect the algorithms which are in either `dotnet.exe`, `hostfxr` or `hostpolicy`. The good news is that `hostfxr` and `hostpolicy` already have JSON parsing capabilities, in fact they already parser `global.json`. The bad news is that for some of these features to work we would need to add JSON parsing capability (and `global.json` parsing) into `dotnet.exe` as well. Technically this is not that hard. The tricky bit is actually updating `dotnet.exe`, so far we've never done that as far as I know.

**Comment by vitek-karas** [comment]
Other related issues:
dotnet/core-setup#3623 - ability to ask for stable-only versions of SDK in global.json
dotnet/core-setup#679 - ability to specify version ranges for SDK in global.json

**Comment by wli3** [comment]
> More ideas, requiring more thought

> [string] Use this --framework as the default for dotnet run -- useful for multi-targeted projects.
> [array of strings] Use these flags for dotnet test by default (like the logger)
> [array of strings] Use these flags for dotnet publish by default (like self-contained of cross-gen)
> [array of strings] Always build for these RIDs
> [array of key/values] Environment variables to set at dotnet run process launch

1. I don't like the `dotnet config` to enable more knobs. We could make existing environment variable to be config. But we should be very conscious for every single new config. Flags does not compose and it will potentially double the test matrix. String and "programmable" configs are even worse. And next step we would need to build tooling around a tool config.

2. Most of the above "config" can be done by customize msbuild in csproj. We should have only one place to do it. If it is hard to do so, we should put effort to make it easy.

**Comment by dasMulli** [comment]
> Most of the above "config" can be done by customize msbuild in csproj

Totally agree. most of these can be MSBuild properties / items and even shared using `Directory.Build.props` though they would need to be advertised through documentation.

Having it inside the project file would also make it available to VS (default run TFM, env vars, publish).

**Comment by KathleenDollard** [comment]
One of the reasons this didn't make the cut is that it makes sense to solve this problem from a broad perspective. This conversation is great because it clarifies some of the complexities...

- global.json is too restrictive and difficult to maintain
- these decisions are often lifestyle choices - not per project choices (what project/user/machine)
- templates, particularly item templates, may need a different strategy
- the git experience of being able to define some things from the command line is nice
- custom package folders should be considered
- specifying the global tools folder should be considered (looks solved by --tool-path)
- specify the RID for the machine (at least when machine doesn't declare)
- consider probing paths
- there are a number of knobs that can be set and we need a strategy programmers can reason about regarding what is in the project file and what is in a config file and what is in global.json. (My personal opinion is that we should hesitate to drive this by the technical need of choosing the EXE).

**Comment by KathleenDollard** [comment]
@richlander I'd like to understand why you said this:

>It happened! We just past that moment where we absolutely need dotnet config. It's time to go build it, I propose for 3.0.

I agree that we need to better tweak SDK selection for preview/not preview to support VS (discussed in other issues) and that we need a long term well thought out strategy for config.

**Comment by nguerrera** [comment]
> (My personal opinion is that we should hesitate to drive this by the technical need of choosing the EXE)

What EXE choice is this exactly?

**Comment by KathleenDollard** [comment]
Bad writing on my part. I edited my comment for clarity.
